src/frequenz/sdk/data_ingestion/resampling/timeseries_resampling_actor.py

```diff
@@ -0,0 +1,159 @@
+        self._output_channels: Dict[TimeSeriesId, Broadcast[Optional[Sample]]] = {}
+        self._resampling_timer = Timer(interval=self._period_s)
+
+    def subscribe(self, subscription: Subscription) -> Receiver[Optional[Sample]]:
+        """Subscribe for data for a specific time series.
+
+        Args:
+            subscription: subscription consisting of `time_series_id`, which is made up
+                of a (component_id, metric_id) pair and `start_time`, which is the
+                point in time from which to subscribe  for data or `None` to subscribe
+                for data from the live component data stream
+
+        Returns:
+            receiver for receiving samples for a specific time series id
+        """
+        if subscription.time_series_id not in self._output_channels:
+            response: SubscriptionResponse = self._data_provider.subscribe(
+                time_series_id=subscription.time_series_id,
+                # Ignore parquet for now [later => start_time=subscription.start_time]
+                start_time=None,
+            )
+            self._input_receivers[subscription.time_series_id] = response.receiver
+            self._output_channels[subscription.time_series_id] = Broadcast[
+                Optional[Sample]
+            ](str(subscription.time_series_id))
+
+        return self._output_channels[subscription.time_series_id].get_receiver()
+
+    async def run(self) -> None:
+        """Run the actor."""
+        while True:
+            component_data_receivers = {
+                str(time_series_id): receiver
+                for time_series_id, receiver in self._input_receivers.items()
+            }
+
+            select = Select(
+                resampling_timer=self._resampling_timer,
+                subscription_receiver=self._subscription_receiver,
+                component_data_receiver=MergeNamed(**component_data_receivers),
+            )
+            while await select.ready():
+                if msg := select.subscription_receiver:
+                    subscription_request: SubscriptionRequest = msg.inner
+                    sample_receiver: Receiver[Optional[Sample]] = self.subscribe(
+                        subscription=subscription_request.subscription
+                    )
+                    await subscription_request.sender.send(
+                        SubscriptionResponse(
+                            receiver=sample_receiver,
+                            subscription=subscription_request.subscription,
+                        )
+                    )
+                    # Breaking out from the loop is required to regenerate
+                    # component_data_receivers to be able to fulfil this
+                    # subscription (later can be optimized by checking if
+                    # an output channel already existed in the `subscribe()` method)
+                    break
```

Comment:
@sahas-subramanian-frequenz what happens when we break the `select.ready()`, I guess all the data that was received by other receivers is still there waiting, right? There is no risk to lose some data? :thinking:

Reply:
Looking at the code of `Select`, I think if there were any other ready receivers in this loop, then the messages will be dropped, so handling a receiver than can end up in a `break` of the loop should be done at the very end to avoid losing messages!

This is also another reason to add support to `Select` to dynamically add/remove receivers, breaking the loop can be very error prone.

Again, to be extra sure, @sahas-subramanian-frequenz maybe you can confirm this?

Reply:
oops,  sorry,  checking now.

Reply:
It looks like upto one message (from any one of the components) could get lost,  and the likelyhood increases as the number of components increases.

Reply:
Yeah, this is a major case IMHO to allow adding/removing receivers from the select loop dynamically, if we really need to use this pattern. I added this info to:

* https://github.com/frequenz-floss/frequenz-channels-python/discussions/24

Reply:
Adding a new comment in a new review to make it more clear, because it wasn't addressed.